### PR TITLE
refactor: extract Moderation PolicyEvaluator

### PR DESCRIPTION
**Pull request by rmusser01**
## Change summary

Pending requester-authored summary. This draft must not be marked ready for review or merged until the requester adds, in their own words, what changed and why these implementation choices were made.

## Summary

- What changed: Extracted Moderation policy decision, scan, snippet, count, and redaction mechanics into a stateless `PolicyEvaluator` with frozen `EvaluationLimits`, while retaining `ModerationService` as the stateful compatibility facade.
- Why: Reduce the service's responsibilities and make policy evaluation independently testable without intentionally changing supported behavior, signatures, dispatch, or exception boundaries.

## Scope

- Production changes are limited to `moderation_service.py` and the new `policy_evaluator.py`.
- No endpoint, schema, shared-model relocation, regex hardening, or intentional behavior changes.
- The branch was transplanted onto latest `dev`; the obsolete shared SDD progress-ledger change was excluded.

## Validation

- [x] Tests added/updated for the preserved behavior and delegation boundaries
- [x] Relevant unit/integration tests pass locally
- [x] Design and implementation documentation included
- [x] Compilation passed for all seven touched Python files
- [x] Black passed for the four new files
- [x] Ruff passed for all touched files; Workflow uses only the documented pre-existing `I001,F401` ignores
- [x] Moderation unit suite: 275 passed
- [x] Endpoint + Guardian: 97 passed
- [x] Chat moderation: 16 passed
- [x] Workflow moderation adapter: 12 passed, 45 deselected
- [x] STT redaction regression: 1 passed
- [x] Bandit: 3,282 LOC, 0 findings, 0 skips, 0 nosec
- [x] `git diff --check origin/dev..HEAD`
- [x] Diff contains exactly the approved 12 files

## UX Audit Checklist

Not applicable: backend-only structural refactor with no WebUI changes.

## Risk & Rollback

- Risk level: Low
- Main risk: subtle drift in legacy Moderation dispatch or redaction edge cases; layered characterization, evaluator, delegation, and caller tests cover those boundaries.
- Rollback plan: revert this PR.

## Merge Gate

- [ ] Requester-authored `Change summary` added above

---
## Summary by cubic
Extracted moderation rule evaluation and redaction into a stateless `PolicyEvaluator` with frozen `EvaluationLimits`, while keeping `ModerationService` as the stateful facade. Behavior is preserved; no endpoint or schema changes.

- **Refactors**
  - Introduced `PolicyEvaluator` and `EvaluationLimits`; `ModerationService` now delegates decision, scan, snippet, count, and redaction.
  - Moved `_UNCATEGORIZED_CATEGORY` to the evaluator and added lazy evaluation caches; removed duplicate logic while preserving public API and exception boundaries.
  - Added characterization, delegation, and integration tests (chat and workflow adapters); included design/plan docs and review remediation notes under `Docs/superpowers/` and backlog tasks.

- **Bug Fixes**
  - Both redaction paths now guard malformed or long-path limits and behave consistently.

<sup>Written for commit 2c9601067b8bdcd67f7daa7bb8eaa1138590e05e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/rmusser01/tldw_server/pull/2770?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: tldw_Server_API/tests/unit/test_moderation_policy_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,1113 @@
...
+
+
+class _LowerUnhashable:
+    def lower(self):
+        return []
+
+
+def test_evaluation_limits_are_frozen():
+    with pytest.raises(FrozenInstanceError):
+        LIMITS.max_scan_chars = 99
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/unit/test_moderation_policy_evaluator.py:87]: Addressed in 9918ba5367. Added the repository-standard module-level pytest.mark.unit marker to all three new evaluator unit-test modules.

File: tldw_Server_API/tests/unit/test_moderation_policy_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,1113 @@
...
+    match_window_chars=5,
+    max_fallback_scan_chars=100,
+    max_replacements_per_pattern=10,
+)
+
+
+def _policy(*rules, **overrides):
+    values = {
+        "enabled": True,
+        "input_enabled": True,
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/tests/unit/test_moderation_policy_evaluator.py]: Addressed in 9918ba5367. Added explicit parameter and return annotations to every non-test helper in the three new evaluator test modules; the AST audit reports no remaining untyped helpers.

File: tldw_Server_API/app/core/Moderation/policy_evaluator.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    TimeoutError,
+    json.JSONDecodeError,
+    re.error,
+)
+
+
+@dataclass(frozen=True)
+class EvaluationLimits:
+    max_scan_chars: int
+    match_window_chars: int
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Moderation/policy_evaluator.py:39]: Addressed in 9918ba5367. Added a class docstring for EvaluationLimits and docstrings for every PolicyEvaluator method; the ModerationService facade docstrings were also restored verbatim from current dev.

File: tldw_Server_API/app/core/Moderation/policy_evaluator.py [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+                if not isinstance(rule, PatternRule):
+                    continue
+                try:
+                    if getattr(rule.regex, "pattern", None) == pattern:
+                        if rule.replacement:
+                            replacement = rule.replacement
+                        break
+                except _EVALUATION_NONCRITICAL_EXCEPTIONS:
+                    continue
+        return self.build_sanitized_snippet_for_replacement(
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/Moderation/policy_evaluator.py:150]: Addressed in 9918ba5367. The best-effort suppression remains behavior-preserving and now includes an explicit rationale: malformed rule metadata must not prevent a sanitized fallback snippet, and this path intentionally does not expose rule details through logging.

**Comment by rmusser01** [comment]
Review remediation is pushed on 2c9601067b. In addition to the four inline findings, the optional policy_types performance item is addressed with a lazy one-entry cache that preserves the staticmethod descriptor and deferred import. Exact-head verification passed 405 focused moderation tests plus the STT policy test, compilation, Black, Ruff, tracking/diff checks, and Bandit with zero findings.
